src/librustc_typeck/collect.rs

```diff
@@ -1299,13 +1300,37 @@ fn trait_def_of_item<'a, 'tcx>(ccx: &CrateCtxt<'a, 'tcx>,
         err.emit();
     }
 
+    let fundamental = {
+        let mut fundamental = None;
+        for attr in tcx.get_attrs(def_id).iter() {
+            if attr.check_name("fundamental") {
+                if fundamental.is_some() {
+                    let mut err = tcx.sess.struct_span_err(
+                        attr.span,
+                        "the #[fundamental] attribute was specified at least twice");
+                    err.emit();
+                    break;
+                }
+
+                if let Some(list) = attr.meta_item_list() {
+                    if attr::list_contains_name(list, "ref_slice") {
+                        fundamental = Some(FundamentalType::SliceRef);
+                    }
+                } else {
+                    fundamental = Some(FundamentalType::All);
+                }
+            }
+        }
+        fundamental.unwrap_or(FundamentalType::None)
+    };
+
     let ty_generics = generics_of_def_id(ccx, def_id);
```

Comment:
@Mark-Simulacrum we can certainly write up an error message. I wouldn't try to pursue generic traits supporting `#[fundamental(sliceref)]` just because this is basically a total hack targeting one narrow use case, not something to be generalized. i.e., this is just a way of saying `&[T]: Trait` is considered 'fundamental'. I suppose we could permit generic traits if we wanted, wouldn't really change too much (the self type would have to be `&[T]`, but the other types would all have to be local, presumably).

Anyway the more interesting question is if we should do this at all =)

@aturon and @withoutboats are giving me second thoughts as to the wisdom of it. I'd feel a lot better if we *had* a plan laid out for negative reasoning and could clearly see that this fits within it (but we don't).